code/modules/reagents/chemistry/machinery/chem_separator.dm

```diff
@@ -0,0 +1,239 @@
+		if(ROUND_UP(reagents.chem_temp - T0C) >= temperature_icon_thresholds[i])
+			threshold = i
+	if(threshold)
+		var/fill_name = "[icon_state]_temp_[temperature_icon_thresholds[threshold]]"
+		var/mutable_appearance/filling = mutable_appearance(icon_state, fill_name)
+		. += filling
+
+/obj/structure/chem_separator/proc/burn_attackby_check(obj/item/I, mob/living/user)
+	var/ignition_message = I.ignition_effect(src, user)
+	if(!ignition_message)
+		return
+	. = TRUE
+	user.visible_message(ignition_message)
+	fire_act(I.get_temperature())
+
+/obj/structure/chem_separator/attackby(obj/item/I, mob/user, params)
+	if(burn_attackby_check(I, user))
+		return
+	if(is_reagent_container(I) && !(I.item_flags & ABSTRACT) && I.is_open_container())
+		var/obj/item/reagent_containers/new_beaker = I
+		if(!user.transferItemToLoc(new_beaker, src))
+			return
+		replace_beaker(user, new_beaker)
+		to_chat(user, span_notice("You add [new_beaker] to [src]."))
+		update_appearance()
+		. = TRUE // no afterattack
+	else
+		return ..()
+
+/obj/structure/chem_separator/proc/replace_beaker(mob/living/user, obj/item/reagent_containers/new_beaker)
+	if(!user)
+		return FALSE
+	if(beaker)
+		if(burning)
+			stop()
+		if(!issilicon(user) && in_range(src, user))
+			user.put_in_hands(beaker)
+		else
+			beaker.forceMove(drop_location())
+		beaker = null
+	if(new_beaker)
+		beaker = new_beaker
+	update_appearance()
+	return TRUE
+
+/obj/structure/chem_separator/fire_act(exposed_temperature, exposed_volume)
+	if(!burning)
+		start()
+		return
+	. = ..()
+
+/obj/structure/chem_separator/extinguish()
+	if(burning)
+		stop()
+	. = ..()
+
+/obj/structure/chem_separator/proc/start()
+	if(!beaker)
+		return
+	if(beaker.reagents.total_volume >= beaker.reagents.maximum_volume)
+		return
+	if(!reagents.total_volume)
+		return
+	separating_reagent = reagents.reagent_list[1].type
+	burning = TRUE
+	update_appearance()
+	START_PROCESSING(SSobj, src)
+
+/obj/structure/chem_separator/proc/stop()
+	separating_reagent = null
+	burning = FALSE
+	update_appearance()
+	STOP_PROCESSING(SSobj, src)
+
+/obj/structure/chem_separator/proc/load()
+	if(burning)
+		return
+	if(!beaker)
+		return
+	if(!beaker.reagents.total_volume)
+		return
+	if(reagents.total_volume >= reagents.maximum_volume)
+		return
+	beaker.reagents.trans_to(reagents, beaker.reagents.total_volume)
+	update_appearance()
+
+/obj/structure/chem_separator/proc/unload()
+	if(burning)
+		return
+	if(!reagents.total_volume)
+		return
+	if(!beaker)
+		return
+	if(beaker.reagents.total_volume >= beaker.reagents.maximum_volume)
+		return
+	reagents.trans_to(beaker.reagents, reagents.total_volume)
+	update_appearance()
+
+/obj/structure/chem_separator/process(delta_time)
+	if(!burning)
+		stop()
+		return
+	if(!beaker)
+		stop()
+		return
+	if(beaker.reagents.total_volume >= beaker.reagents.maximum_volume)
+		stop()
+		return
+	if(!reagents.get_reagent_amount(separating_reagent))
+		stop()
+		return
```

Comment:
I think all of these should go into a should_process proc so that we can just call stop() from one place.

Reply:
I don't get it. Do you want to have a separate procs for all guard clauses? Like should_process(), should_load(), should_unload()? What is the point if they're not reused?

Reply:
try combining if clauses at least, I'm only not sure if splitting the line like this is syntactically correct, if not you can add ` \` before newlines.

```suggestion
	if(!burning || !beaker
			|| beaker.reagents.total_volume >= beaker.reagents.maximum_volume
			|| !reagents.get_reagent_amount(separating_reagent))
		stop()
		return
```

Reply:
And I think this expression itself can be reused in many places, if you combine it like that, hence why mothblocks proposed a separate proc.

Reply:
Yeah something like that or a single proc that contains all of them and does `return FALSE`, which you then call inside here, and stop() if it fails.

Reply:
The point is to stop you from copying and pasting stop() everywhere, and so we can change that if necessary more easily later